// ncloudchat/lib/CoreManager.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports["default"] = void 0;

/**
 * Copyright (c) NBASE CORP. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var config = {
  // Defaults
  IS_NODE: typeof process !== "undefined" && !!process.versions && !!process.versions.node,
  REQUEST_ATTEMPT_LIMIT: 5,
  REQUEST_BATCH_SIZE: 20,
  REQUEST_HEADERS: {},
  TOKEN: "",
  SERVER_URL: "https://ncloudchat.apigw.ntruss.com/gpapps/v1",
  SOCKET_URL: "https://soc.ncloudchat.ntruss.com",
  // jp / sg
  // SERVER_URL: "https://ncloudchat.apigw.ntruss.com/gpapps/jp-v1",
  // SOCKET_URL: "https://jp-soc.ncloudchat.ntruss.com",
  VERSION: "js" + require("../package.json").version,
  PROJECT_ID: null
};
var CoreManager = {
  get: function get(key) {
    if (config.hasOwnProperty(key)) {
      return config[key];
    }

    throw new Error("Configuration key not found: " + key);
  },
  set: function set(key, value) {
    config[key] = value;
  }
};
var _default = CoreManager;
exports["default"] = _default;
