// ncloudchat/lib/Type.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.MessageType = exports.ChannelType = void 0;

/**
 * Copyright (c) NBASE CORP. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

/**
 * Enum for channel types.
 * @readonly
 * @enum {string}
 */
var ChannelType;
exports.ChannelType = ChannelType;

(function (ChannelType) {
  ChannelType["PRIVATE"] = "PRIVATE";
  ChannelType["PUBLIC"] = "PUBLIC";
})(ChannelType || (exports.ChannelType = ChannelType = {}));
/**
 * Enum for message types.
 * @readonly
 * @enum {string}
 */


var MessageType;
exports.MessageType = MessageType;

(function (MessageType) {
  MessageType["text"] = "text";
  MessageType["photo"] = "photo";
  MessageType["file"] = "file";
})(MessageType || (exports.MessageType = MessageType = {}));
