// ncloudchat/esm/mutations/message.js
/**
 * Copyright (c) NBASE CORP. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import CoreManager from "../CoreManager";
import { fetchData } from "../Network";
import { translateQuery, messageQuery } from "../graphql/message";
/**
 * Trasnslate a text block to the target language from the source language.
 *
 * @async
 * @function translate
 * @param {string} channelId - The id of a channel.
 * @param {string} srcLang - The source language.
 * @param {string} targetLang - The target language.
 * @param {string} text - The text block to be translated.
 * @returns {Promise<any>}
 */
export var translate = function (channelId, srcLang, targetLang, text) { return __awaiter(void 0, void 0, void 0, function () {
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, fetchData("", translateQuery, {
                    projectId: CoreManager.get("PROJECT_ID"),
                    channelId: channelId,
                    srcLang: srcLang,
                    targetLang: targetLang,
                    text: text
                })];
            case 1: return [2 /*return*/, _a.sent()];
        }
    });
}); };
/**
 * Upload a file to a channel. Only supports png, jpg, jpeg file types.
 *
 * @async
 * @function upload
 * @param {string} channelId - The id of a channel.
 * @param {any} file - An image file.
 * @returns {Promise<any>}
 */
export var upload = function (channelId, file) { return __awaiter(void 0, void 0, void 0, function () {
    var token, endpoint, form, result;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                token = CoreManager.get("TOKEN");
                if (!token) {
                    console.error("Token is not found");
                    return [2 /*return*/];
                }
                endpoint = CoreManager.get("SERVER_URL") + "/upload";
                form = new FormData();
                form.append("file", file);
                form.append("channelId", channelId);
                form.append("upload", "1");
                return [4 /*yield*/, fetch(endpoint, {
                        method: "POST",
                        headers: {
                            Authorization: "Bearer " + token,
                            "X-PROJECT-ID": CoreManager.get("PROJECT_ID"),
                            projectid: CoreManager.get("PROJECT_ID")
                        },
                        body: form
                    }).then(function (res) { return res.json(); })];
            case 1:
                result = _a.sent();
                return [2 /*return*/, result];
        }
    });
}); };
/**
 * Send a Message
 *
 * @async
 * @function translate
 * @param {string} data - The id of a message.
 * @returns {Promise<any>}
 */
export var message = function (data) { return __awaiter(void 0, void 0, void 0, function () {
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, fetchData("createChannelMessage", messageQuery, {
                    projectId: CoreManager.get("PROJECT_ID"),
                    channelId: data.channelId,
                    content: data.message
                })];
            case 1: return [2 /*return*/, _a.sent()];
        }
    });
}); };
