// ncloudchat/esm/CoreManager.js
/**
 * Copyright (c) NBASE CORP. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var config = {
    // Defaults
    IS_NODE: typeof process !== "undefined" &&
        !!process.versions &&
        !!process.versions.node,
    REQUEST_ATTEMPT_LIMIT: 5,
    REQUEST_BATCH_SIZE: 20,
    REQUEST_HEADERS: {},
    TOKEN: "",
    SERVER_URL: "https://ncloudchat.apigw.ntruss.com/gpapps/v1",
    SOCKET_URL: "https://soc.ncloudchat.ntruss.com",
    // jp / sg
    // SERVER_URL: "https://ncloudchat.apigw.ntruss.com/gpapps/jp-v1",
    // SOCKET_URL: "https://jp-soc.ncloudchat.ntruss.com",
    VERSION: "js" + require("../package.json").version,
    PROJECT_ID: null
};
var CoreManager = {
    get: function (key) {
        if (config.hasOwnProperty(key)) {
            return config[key];
        }
        throw new Error("Configuration key not found: " + key);
    },
    set: function (key, value) {
        config[key] = value;
    }
};
export default CoreManager;
